static_any fixes

thirtythreeforty opened the pull request:
This fixes the alignment issues [I mentioned on Reddit](https://www.reddit.com/r/cpp/comments/727sgr/inplace_containers_for_fun_and_profit/dngx0na/) by using `std::aligned_storage_t` as backing storage.  It also overhauls the copying code (in a couple cases, the container would memcpy live objects, something you can obviously only do with POD types).  Please review my changes for strong exception guarantee violations; there shouldn't be any but that's what review is for.

There are also a couple of other minor fixes.

I had to remove the unit test checking the size of `static_any`; the size of `aligned_storage_t` is not guaranteed as far as I know.  If this is a deal-breaker, let me know; I don't care about the exact size of the container, but if you're counting bytes to fit in a cache line, you might.

thirtythreeforty committed:
Inline bad_any_cast ctor and dtor's

thirtythreeforty committed:
Use aligned_storage_t for all raw storage

thirtythreeforty committed:
Soundness fix: Do not memcpy live, non-POD objects

thirtythreeforty committed:
When assigning from empty, actually destroy our object

thirtythreeforty committed:
Only use temporary buffers for throwing operations

thirtythreeforty commented:
Benchmarks seems similar after the copy code overhaul, but please double-check me on those too to make sure I haven't regressed.

thirtythreeforty committed:
Fix size of static_any in readme example

The old code Works on My Machine™ but not on Travis.

Normally I'd use std::max(sizeof(all_args)...) but since this is a
readme example I think the hardcoded number should be clear enough and
is large enough for most platforms.

thirtythreeforty commented on any.hpp:
```diff
@@ -221,20 +228,32 @@ struct static_any
+			old_function = function_;
+		}
+
+		destroy();
```
Hold off on merging this just yet, I need to rethink this implementation.

One issue I just thought of is that throwing move constructors (written by insane users) will cause interesting problems in `assign`.  Everything is fine until the call to `call_copy_or_move`; what we'd really like is to call_move_if_noexcept.

More generally, making a temporary copy of the storage in 239 is invalid, because we destroy it in line 243.

Lemme think about this.

thirtythreeforty replied:
I think the rule of thumb is that you have to abandon strong exception guarantees if the contained type does not have a nothrow copy or move constructor.

david-grs commented:
Hey! Thanks for the fixes. I didnt have a look yet but will definitely soon - I need a few days.

thirtythreeforty commented:
That's fine, no rush!

thirtythreeforty committed:
Implement move-assignment from static_any

thirtythreeforty committed:
Unit tests: Replace explicit sizes with portable calculations

This should fix MSVC's problems.

thirtythreeforty commented:
Well I tried.

thirtythreeforty committed:
Make CallCounter also count alive instances

This exposes a double-free made by static_any in the
any_exception::copy_from_any test case.

thirtythreeforty commented:
Okay, so I think the reason I'm having trouble here is that I believe you can't simultaneously guarantee all of:

- strong exception safety
- no dynamic allocations

...for *all* types.

The problem is illustrated by the commit I just made.  `CallCounter` exposes a double-free made by static_any (the first free occurs on `unit_tests.cpp:581`, and the second is made when `b` is destroyed).

What you *can* provide is the strong exception guarantee *only if* the currently-stored type has a noexcept move or copy constructor (allowing you to use the second-buffer-on-the-stack trick).  If the currently-stored type is not noexcept, the container has to fall back to the basic exception guarantee.  In my opinion, this is a sane guarantee to make, but it is weaker than the guarantee made by the readme.

I can fix the code to have those semantics, but I just wanted to make sure you're on board with the change.
